File: hub/src/test/unit/com/thoughtworks/selenium/grid/hub/hub_servlet_test.rb
require File.expand_path(File.dirname(__FILE__) + "/../../../../../unit_test_helper")

include_class 'com.thoughtworks.selenium.grid.hub.HubServlet'

class HubServletTest < Test::Unit::TestCase

  xtest "jruby broken - Clint", "replay set content type as plain text" do
    remote_control_response = Response.new 123, ""
    servlet_response = HttpServletResponse.new
    servlet_response.expects("setContentType").with("text/plain")
    servlet_response.expects("getWriter").returns PrintWriter.new(StringWriter.new(100))

    HubServlet.new.reply(servlet_response, remote_control_response)
  end

  xtest "jruby broken - Clint", "reply set character encoding to UTF8" do
    remote_control_response = Response.new 123, ""
    servlet_response = HttpServletResponse.new
    servlet_response.expects("setCharacterEncoding").with("UTF-8")
    servlet_response.expects("getWriter").returns PrintWriter.new(StringWriter.new(100))

    HubServlet.new.reply(servlet_response, remote_control_response)
  end

  xtest "jruby broken - Clint", "reply set status from remote control on the servlet response" do
    remote_control_response = Response.new 123, ""
    servlet_response = HttpServletResponse.new
    servlet_response.expects("setStatus").with 123
    servlet_response.expects("getWriter").returns PrintWriter.new(StringWriter.new(100))
  end

  xtest "jruby broken - Clint", "reply write remote control response on servlet response as plain text" do
    writer = StringWriter.new 100
    remote_control_response = Response.new 0, "some response message"
    servlet_response = HttpServletResponse.new
    servlet_response.expects("getWriter").returns PrintWriter.new(writer)
    HubServlet.new.reply(servlet_response, remote_control_response)

    assert_equal "some response message", writer.get_buffer.to_string
  end

  xtest "jruby broken - Clint", "forward execute the selenese command on the appropriate remote control" do
    servlet = HubServlet.new
    request_parameters = ServletParametersAdapter.new
    request_parameters.put "cmd", "aSeleneseCommand"
    request_parameters.put "sessionId", "a session id"
    pool = DynamicRemoteControlPool.new
    remote_control = RemoteControlProxy.new
    environment_manager = EnvironmentManager.new
    response = Response.new 0, ""

    pool.expects("retrieve").with("a session id").returns remote_control
    remote_control.expects("forward").with(eq("cmd=aSeleneseCommand&sessionId=a+session+id")).returns response
    assert_equal response, servlet.forward(request_parameters, pool, environment_manager)
  end

  xtest "jruby broken - Clint", "forward returns an error message when a command parsing exception is thrown" do
    servlet = HubServlet.new
    request_parameters = ServletParamatersAdapter.new
    request_parameters.put "cmd", "aSeleneseCommand"
    request_parameters.put "sessionId", "a session id"
    pool = DynamicRemoteControlPool.new
    environment_manager = EnvironmentManager.new

    pool.expects("retrieve").with("a session id").throws CommandParsingException.new("an error message")
    response = servlet.forward(request_parameters, pool, environment_manager)
    assert_equal 200, response.status_code
    assert_equal "ERROR: an error message", response.body
  end

  xtest "jruby broken - Clint", "request parameters returns adapted request parameters" do
    parameter_map = new HashMap#<String, String[]>
    parameter_map.put "a name", ["a value"]
    request = HttpServletRequest.new
    request.expects("getParameterMap").returns parameter_map
    servlet = HubServlet.new
    parameters = servlet.request_parameters request
    assert_equal "a value", parameters.get("a name")
  end

end

File: hub/src/test/unit/java/com/thoughtworks/selenium/grid/hub/HubServletTest.java
package com.thoughtworks.selenium.grid.hub;

import com.thoughtworks.selenium.grid.hub.remotecontrol.DynamicRemoteControlPool;
import com.thoughtworks.selenium.grid.hub.remotecontrol.RemoteControlProxy;
import com.thoughtworks.selenium.grid.Response;
import com.thoughtworks.selenium.grid.HttpParameters;
import static junit.framework.Assert.assertEquals;
import org.jbehave.classmock.UsingClassMock;
import org.jbehave.core.mock.Mock;
import org.junit.Test;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.io.PrintWriter;
import java.io.StringWriter;
import java.util.HashMap;
import java.util.Map;


public class HubServletTest extends UsingClassMock {
    
    @Test
    public void replySetContentTypeAsPlainText() throws IOException {
        final Response remoteControlResponse;
        final Mock servletResponse;

        remoteControlResponse = new Response(123, "");
        servletResponse = mock(HttpServletResponse.class);
        servletResponse.expects("setContentType").with("text/plain");
        servletResponse.expects("getWriter").will(returnValue(new PrintWriter(new StringWriter(100))));

        new HubServlet().reply((HttpServletResponse) servletResponse, remoteControlResponse);
        verifyMocks();
    }

    @Test
    public void replySetCharacterEncodingToUTF8() throws IOException {
        final Response remoteControlResponse;
        final Mock servletResponse;

        remoteControlResponse = new Response(123, "");
        servletResponse = mock(HttpServletResponse.class);
        servletResponse.expects("setCharacterEncoding").with("UTF-8");
        servletResponse.expects("getWriter").will(returnValue(new PrintWriter(new StringWriter(100))));

        new HubServlet().reply((HttpServletResponse) servletResponse, remoteControlResponse);
        verifyMocks();
    }

    @Test
    public void replySetStatusFromRemoteControlOnTheServletResponse() throws IOException {
        final Response remoteControlResponse;
        final Mock servletResponse;

        remoteControlResponse = new Response(123, "");
        servletResponse = mock(HttpServletResponse.class);
        servletResponse.expects("setStatus").with(123);
        servletResponse.expects("getWriter").will(returnValue(new PrintWriter(new StringWriter(100))));

        new HubServlet().reply((HttpServletResponse) servletResponse, remoteControlResponse);
        verifyMocks();
    }

    @Test
    public void replyWriteRemoteControlResponseOnServletResponseAsPlainText() throws IOException {
        final StringWriter writer = new StringWriter(100);
        final Response remoteControlResponse;
        final Mock servletResponse;

        remoteControlResponse = new Response(0, "some response message");
        servletResponse = mock(HttpServletResponse.class);
        servletResponse.expects("getWriter").will(returnValue(new PrintWriter(writer)));

        new HubServlet().reply((HttpServletResponse) servletResponse, remoteControlResponse);
        assertEquals("some response message", writer.getBuffer().toString());

        verifyMocks();
    }

    @Test
    public void forwardExecuteTheSeleneseCommandOnTheAppropriateRemoteControl() throws IOException, NoSuchMethodException {
        final HttpParameters requestParameters;
        final Mock environmentManager;
        final Mock remoteControl;
        final Response response;
        final HubServlet servlet;
        final Mock pool;

        servlet = new HubServlet();
        requestParameters = new HttpParameters();
        requestParameters.put("cmd", "aSeleneseCommand");
        requestParameters.put("sessionId", "a session id");
        pool = mock(DynamicRemoteControlPool.class);
        remoteControl = mock(RemoteControlProxy.class);
        environmentManager = mock(EnvironmentManager.class);
        response = new Response(0, "");

        pool.expects("retrieve").with("a session id").will(returnValue(remoteControl));
        remoteControl.expects("forward").with(eq(requestParameters)).will(returnValue(response));

        assertEquals(response, servlet.forward(requestParameters, (DynamicRemoteControlPool) pool, (EnvironmentManager) environmentManager));
        verifyMocks();
    }

    @SuppressWarnings({"ThrowableInstanceNeverThrown"})
    @Test
    public void forwardReturnAnErrorMessageWhenACommandParsingExceptionIsThrown() throws IOException, NoSuchMethodException {
        final HttpParameters requestParameters;
        final Mock environmentManager;
        final Response response;
        final HubServlet servlet;
        final Mock pool;

        servlet = new HubServlet();
        requestParameters = new HttpParameters();
        requestParameters.put("cmd", "aSeleneseCommand");
        requestParameters.put("sessionId", "a session id");
        pool = mock(DynamicRemoteControlPool.class);
        environmentManager = mock(EnvironmentManager.class);

        pool.expects("retrieve").with("a session id").will(throwException(new CommandParsingException("an error message")));

        response = servlet.forward(requestParameters, (DynamicRemoteControlPool) pool, (EnvironmentManager) environmentManager);
        assertEquals(200, response.statusCode());
        assertEquals("ERROR: an error message", response.body());
        verifyMocks();
    }

    @Test
    public void requestParametersReturnsAdaptedRequestParameters() {
        final HttpParameters parameters;
        final Map<String, String[]> parameterMap;
        final HubServlet servlet;
        final Mock request;

        parameterMap = new HashMap<String, String[]>();
        parameterMap.put("a name", new String[] { "a value"});
        request = mock(HttpServletRequest.class);
        request.expects("getParameterMap").will(returnValue(parameterMap));
        servlet = new HubServlet();
        parameters = servlet.requestParameters((HttpServletRequest) request);
        assertEquals("a value", parameters.get("a name"));
    }


}
